chains/solana/contracts/programs/ccip-offramp/src/instructions/v1/buffering.rs

```diff
@@ -0,0 +1,328 @@
+            require_eq!(
+                chunk_index,
+                self.total_chunks - 1,
+                CcipOfframpError::ExecutionReportBufferInvalidChunkSize
+            );
+        }
+
+        require_gt!(
+            self.total_chunks,
+            chunk_index,
+            CcipOfframpError::ExecutionReportBufferInvalidChunkIndex
+        );
+
+        let start = self.chunk_length as usize * chunk_index as usize;
+        let end = self.data.len().min(start + chunk.len());
+        self.data[start..end].copy_from_slice(chunk);
+        self.chunk_bitmap |= chunk_mask;
+
+        Ok(())
+    }
+}
+
+impl ExecutionReportBuffer {
+    fn initialize(
+        &mut self,
+        report_length: u32,
+        chunk_length: u32,
+        total_chunks: u8,
+        chunk_index: u8,
+    ) -> Result<()> {
+        require!(
+            !self.is_initialized(),
+            CcipOfframpError::ExecutionReportBufferAlreadyInitialized
+        );
+        require!(
+            report_length > 0 && chunk_length <= report_length && chunk_length > 0,
+            CcipOfframpError::ExecutionReportBufferInvalidLength
+        );
+        require!(
+            total_chunks < 64 && total_chunks > 0,
+            CcipOfframpError::ExecutionReportBufferInvalidChunkSize
+        );
+
+        // If we're initializing with the last chunk, it could be smaller
+        // than the rest, so we calculate the chunk size based on the expected
+        // size of all the others.
+        let is_last = chunk_index == total_chunks - 1;
+        let global_chunk_length = if is_last && total_chunks > 1 {
+            (report_length - chunk_length) / (total_chunks as u32 - 1)
+        } else {
+            chunk_length
+        };
+
+        require_eq!(
+            total_chunks as u32,
+            div_ceil(report_length, global_chunk_length),
+            CcipOfframpError::ExecutionReportBufferInvalidLength,
+        );
+        self.chunk_length = global_chunk_length;
+        self.total_chunks = total_chunks;
+        self.data.resize(report_length as usize, 0);
+
+        Ok(())
+    }
+}
+
+pub fn deserialize_from_buffer_account(
+    execution_report_buffer: &AccountInfo,
+) -> Result<(ExecutionReportSingleChain, usize)> {
+    // Ensures the buffer is initialized, and owned by the program.
+    require_keys_eq!(
+        *execution_report_buffer.owner,
+        crate::ID,
+        CcipOfframpError::ExecutionReportUnavailable
+    );
+    let buffer = ExecutionReportBuffer::try_deserialize(
+        &mut execution_report_buffer.data.borrow().as_ref(),
+    )?;
+
+    Ok((
+        ExecutionReportSingleChain::deserialize(&mut buffer.bytes()?)
+            .map_err(|_| CcipOfframpError::FailedToDeserializeReport)?,
+        buffer.data.len(),
+    ))
+}
+
+fn div_ceil<T: Into<u32>>(a: T, b: T) -> u32 {
+    let (a, b) = (a.into(), b.into());
+    assert!(a + b - 1 > 0);
+    (a + b - 1) / b
+}
```

Comment:
Qq, is there any reason to not use `a.div_ceil(b)` directly, given that `u32` implements that method?

Reply:
It's not stable on the Rustc version we use on CI I believe.